### PR TITLE
cli: introduce role-based lint configuration setup

### PR DESCRIPTION
**Pull request by Rugvip**
## Hey, I just made a Pull Request!

Tried a lot of different ways of doing this, including separate configuration files for different roles, and a root configuration that applied configuration to all packages. This is where it ended up in the end 😅.

One feature of the ESLint configuration I really wanted to add are the [helper options](https://github.com/backstage/backstage/blob/843d77d7639c3360a26feaa3d3577405ea0bc600/packages/cli/config/eslint-factory.js#L24-L31) for defining additional syntax and import rules in different contexts. Right now that can get [really messy](https://github.com/backstage/backstage/blob/c8f5f5c1d24f254bbc57848dd082b6a394d5243e/plugins/scaffolder-backend/.eslintrc.js#L10-L23).

This ofc also moves over to a single lint entry point for all packages, with the exact configuration being selected based on the package role. The default export of the entry point is the one that detects it based on the package, but it's possible to use `createConfigForRole` directly if one wants to hardcode the role somewhere.

I'm also introducing a `backstage-cli migrate package-lint-configs`, to go with the other role migration commands. Still kept the changeset lightweight on migration details, but I intend to start working on that piece now :grin:

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/scaffolder-backend/.eslintrc.js [hunk 1]
```diff
@@ -1,41 +1,18 @@
-const parent = require('@backstage/cli/config/eslint.backend');
-
-module.exports = {
-  extends: [require.resolve('@backstage/cli/config/eslint.backend')],
+module.exports = require('@backstage/cli/config/eslint-factory')(__dirname, {
```
Comment (jhaals) [inline plugins/scaffolder-backend/.eslintrc.js:1]: This looks a lot cleaner! 👍 

In a future world when I want all the default rules except for X, is there a way of doing that our do I have to opt out?

File: .changeset/gentle-dancers-greet.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+---
+'@backstage/cli': patch
+---
+
+Added a new ESLint configuration setup for packages, which utilizes package roles to generate the correct configuration. The new configuration is available at `@backstage/cli/config/eslint-factory`.
+
+Introduced a new `backstage-cli migrate package-lint-configs` command, which migrates old lint configurations to use `@backstage/cli/config/eslint-factory`.
```
Comment (jhaals) [inline .changeset/gentle-dancers-greet.md:7]: Is this something we should recommend that integrators run for their own plugins?

Comment (jhaals) [reply .changeset/gentle-dancers-greet.md:7]: Realise that the PR body mentioned migration as the next step so I believe that answers that :)

**Review by jhaals — APPROVED** [review]
Nice work!

File: plugins/scaffolder-backend/.eslintrc.js (hunk 1 above)
Comment (Rugvip) [reply plugins/scaffolder-backend/.eslintrc.js:1]: Yep you can disable rules by setting them to `'off'` or `0`

Comment (Rugvip) [reply plugins/scaffolder-backend/.eslintrc.js:1]: Realize that if you meant the restricted imports etc. then it's only possible to add atm. You can completely replace the rule configuration as well though

**Review by freben — APPROVED** [review]
nnnneat

**Pull request merged**